docs/RFCS/20220711_external_connections.md

```diff
@@ -0,0 +1,407 @@
+they likely access the external connection more than once during their execution. To enable this,
+every job persists its URI in the `system.jobs` table. Every job is individually responsible for
+redacting sensitive information, before displaying the persisted URI via the various job
+observability tools that CockroachDB offers. There have been several occasions where we have leaked
+secrets because a job had forgotten to sanitize the URI before displaying it.
+
+At the root of these shortcomings is the tight coupling between the systems that use the external
+connection, and the representation of the external connection. Every operation that accesses an
+external connection is also made aware of the raw URI that is parsed to configure, authenticate, and
+interact with the connection. This RFC proposes separating the creation and management of an
+External Connection object from its use. This separation would facilitate the development of a
+standardized interface and a fine-grained permission model, which would create a more user-friendly,
+secure experience.
+
+## Technical design
+
+### Creating an External Connection
+
+An External Connection object represents a resource that resides outside of CockroachDB. In the
+scope of this RFC, the discussion is limited to the two types of external connections that are used
+by bulk operations and changefeeds, namely:
+
+- Storage: a connection capable of reading and writing bytes to an external store.
+
+- Key Management Service (KMS): a connection capable of interacting with keys hosted in a KMS, to
+  encrypt and decrypt data.
+
+External Connections will be persisted in a new system table called `system.external_connections`
+with the following schema:
+
+```sql
+CREATE TABLE system.external_connections (
+    name               STRING NOT NULL,
+    created            TIMESTAMPZ NOT NULL DEFAULT NOW(),
+    connection_type    STRING NOT NULL,
+    connection_details BYTES,
+
+    CONSTRAINT "primary" PRIMARY KEY name,
+)
+```
+
+The following SQL syntax will be used to create an External Connection:
+
+```sql
+CREATE EXTERNAL CONNECTION <name> AS <scheme://uri.....?params> [WITH connection_options]
+```
+
+An External Connection object is uniquely identifiable by its `name`, which is also how it will be
+referred to by other database operations, for example, backup and changefeeds. `name` is a free-form
+user input string that is specified at the time of creation. The creation of an External Connection
+is a privileged operation as discussed in the privilege model below.
+
+The `created` column stores the timestamp at which the External Connection was created.
+
+The `connection_type` categorizes the External Connection based on the capabilities of the resource
+it represents. To begin with, we will support the `Storage` and `KMS` connection types, each
+corresponding to a unique connection capability as described above. Every connection type can in
```

Comment:
Worth commenting on why we don't have a type specific to changefeed sinks (such as kafka, gcppubsub, etc) that can't be used with backups?